pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -122,6 +122,59 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 
+	return ValidateObjectParameters(paramSpecs, params)
+}
+
+// ValidateObjectParameters checks if all required keys of object type params are provided in taskrun params.
+// To use this function externally, make sure `paramSpecs` and `params` have
+// been checked against general missing/extra params and param type alignment,
+// which been done in `validateParams` in this module.
+func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
+	neededKeys := make(map[string][]string)
+	providedKeys := make(map[string][]string)
+
+	// collect needed keys for object parameters
+	for _, spec := range paramSpecs {
+		if spec.Type == v1beta1.ParamTypeObject {
+			for key := range spec.Properties {
+				neededKeys[spec.Name] = append(neededKeys[spec.Name], key)
+			}
+		}
+	}
+
+	// collect provided keys for object parameters
+	for _, p := range params {
+		if p.Value.Type == v1beta1.ParamTypeObject {
+			for key := range p.Value.ObjectVal {
+				providedKeys[p.Name] = append(providedKeys[p.Name], key)
+			}
+		}
+	}
+
+	return ValidateObjectKeysNeededAgainstProvided(neededKeys, providedKeys)
+}
+
+// ValidateObjectKeysNeededAgainstProvided checks if objects have missing keys in its provider (either taskrun value or result value)
+// To use this function at result side, make sure `providedObjectParamKeys` and `neededObjectParamKeys` have
+// been checked against missing/extra object which been done in `validateParams` in this module.
+func ValidateObjectKeysNeededAgainstProvided(neededObjectKeys, providedObjectKeys map[string][]string) error {
+	missings := []string{}
+	for p, keys := range providedObjectKeys {
+		if _, ok := neededObjectKeys[p]; !ok {
+			// Ignore any missing objects - this happens when extra objects were
+			// passed that aren't being used.
+			continue
+		}
+		missedKeys := list.DiffLeft(neededObjectKeys[p], keys)
```

Comment:
This seems to be a O(n^2) function and not so efficient .

Reply:
This `list.diffleft` is existing implemented function that has been used in [multiple places](https://github.com/tektoncd/pipeline/search?q=list.diffleft) to compare needed vs provided.

Reply:
I see why we use it, it is a good way to find out all missing keys, also make sense

Reply:
But this func is not implemented time efficient, maybe we want to save more memory...

Reply:
> But this func is not implemented time efficient, maybe we want to save more memory...

are you referring to `list.diffleft` or something else?

Reply:
yes

Reply:
i see. that's a pre-existing utility function in the code base. Let's look into it later.